### ISSUE TITLE
Same url for endpoint details and endpoint message list

### ISSUE THREAD

**Issue by koriaf** [ausdigital/ausdigital-tap-gw#3]
The ausdigital tap-gw spec (http://ausdigital.org/specs/ausdigital-tap-gw/1.0/) says:

> GET endpoint details GET /endpoints/{uuid}

but

> get list of new messages with GET /endpoints/{uuid}/{filter}

First, I think we shall separate filters from url by `?` character, making them GET parameters.
Second, I think that list of new messages shall be retrieved by new `/messages/` endpoint, assuming:
* user is authenticated - so it's easy for tap-gw to filter messages by given user
* this endpoint may support `endpoint=XXX` GET parameter to effectively filter messages by endpoints (if user wants it)
* messages may be filtered by `read=true/false` parameter as well

**Comment by monkeypants** [comment]
> get list of new messages with GET /endpoints/{uuid}/{filter}

That's obviously a mistake/wrong. It should at least be `GET /endpoints/{uuid}/messages?{filter}`. Are you suggesting `GET /messages/?{filter}` and `GET /messages/{uuid}`? That's a fair suggestion. Is that what you meant?

**Comment by koriaf** [comment]
Yes, I think we don't need to add endpoint ID to message list URL because it forces user to filter by endpoint. We shall use `/messages/?{optiona filter}` so user may just receive all messages he has by single request, and filter them if it's required.

**Comment by monkeypants** [comment]
I'm thinking about a ledger or access point servicing a large number of participants. They will always apply customer filter. For a ledger servicing a small number of participants, they might actually want all messages of a particular status (batch download etc).

One reason to preference `GET /endpoints/{uuid}/messages?{filter}` is that HTTP infrastructure could load balance to parallel/sharded implementations. For example, HTTP router could reverse proxy based on `{endpoint}`.  I know our testpoint implementation has a stateless web tier with distributed persistence tier, but that might not be the case for all integrations on the network.

If the endpoint is specified in the filter to the right of the `?`, that means the web tier does not have the opportunity to route to shards, so which places constraints on implementers. Keeping the `/endpoints/{uuid}/` in the path might be convenient for some implementations.

I don't know how important this is. Apart from a cleaner URL, what is the advantage of providing a `messages` resource that spans multiple endpoints?

**Comment by koriaf** [comment]
* it's theoretically possible to route requests based on GET parameters (at least nginx supports it)
* it's theoretically possible for some implementations to support `/endpoints/{eid}/messages/` transparently (or not transparently) by http rewrite functionality (change whatever custom url they support to any standard url).

I think it's implementation prerogative to make whatever url they consider useful and say their customers "Feel free to use this url for your convenience or standard ausdigital one for strict protocol support".

My belief is that specification shall be as simple as possible, and /messages/ is obviously simpler :-) What does everyone else think? May be someone already has some ledger implementation experience and can say their preference.

**Comment by kshychko** [comment]
as we include an endpoint as a message attribute in a response for this request:
```
"data": [
      {
          "type": "TapMessage",
          "id": "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXX",
          "attributes":{
              "uuid": "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXX",
              "status": "sent",
              "createdAt": "2017-02-27T05:52:27.618962Z",
              "sentAt": "2017-02-27T05:52:46.081878Z",
              "docId": "QmXXXXXXXXXXXXXXXXXXXXXXXXXXX",
              "status": "new",
              "metadata":{
                  "key1": "value2",
                  "key2": "value2"
              },
              "endpoint": "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXX"
          }
      }
  ]
```
For me it seems more reasonable to have it as a query parameter (the same as status) instead of a path parameter.

**Comment by monkeypants** [comment]
OK, I am persuaded that GET parameter is better than path parameter.

**Comment by kshychko** [comment]
TAP-GW swagger document has been updated. 
Please check query parameters for Get TAP Messages:
http://ausdigital.org/specs/ausdigital-tap-gw/1.0/api#getget-tap-messages

**Comment by koriaf** [comment]
Now I'm happy about the swagger, just need to reflect changes in the spec.
